src/rust/ide/view/graph-editor/src/lib.rs

```diff
@@ -937,6 +967,77 @@ impl Edges {
 
 
 
+// =========================
+// === ExecutionStatuses ===
+// =========================
+
+/// [`ExecutionsStatuses`] can be used to collect the execution statuses of all nodes in a graph. It
+/// exposes their minimum and maximum running times through its FRP endpoints. The structure needs
+/// to be updated whenever a node is added or deleted or changes its execution status.
+mod execution_statuses {
+    use super::*;
+    use std::cmp::Ordering::Equal;
+
+    ensogl::define_endpoints! {
+        Output {
+            min_duration (f32),
+            max_duration (f32),
+        }
+    }
+
+    #[derive(Debug,Clone,CloneRef,Default)]
+    pub struct ExecutionStatuses {
+        frp       : Frp,
+        // If performance becomes an issue then we could try to store the durations in efficient
+        // priority queues instead, one to determine the minimum and one to determine the maximum.
+        durations : SharedHashMap<NodeId,f32>
+    }
+
+    impl Deref for ExecutionStatuses {
+        type Target = Frp;
+
+        fn deref(&self) -> &Self::Target {
+            &self.frp
+        }
+    }
+
+    impl ExecutionStatuses {
+        pub fn new() -> Self {
+            let frp       = Frp::new();
+            let durations = SharedHashMap::new();
+            Self {frp,durations}
+        }
+
+        pub fn set(&self,node:NodeId,status:NodeExecutionStatus) {
+            match status {
+                NodeExecutionStatus::Finished {duration} => self.durations.insert(node,duration),
+                _ => self.durations.remove(&node),
+            };
+            self.update_min_max();
+        }
+
+        pub fn remove(&self,node:NodeId) {
+            self.durations.remove(&node);
+            self.update_min_max();
+        }
```

Comment:
This API and logic can also be done in FRP.

Reply:
I would do it this way:

1. in define_endpoints macro above add inputs "set" and "remove" with proper inputs
2. in add FRP nodes mapping the "set" and "remove" inputs, so they change `durations` set accordingly and return the min/max times.
3. redirect results from nodes defined in 2 to the outputs.

Then, instead of emitting `set` you can redirect some other FRP output with `<+`

And yes, I think it could be profitable; some time ago we had the debug printer for FRP graphs, which we may restore some day. And then we will see the relationship between calling set and updaing min/max time.